larsim/LegacyLArG4/ISCalculationCorrelated.cxx

```diff
@@ -0,0 +1,153 @@
+//            of hard-coding it into this algorithm
+//
+// \author  wforeman @ iit.edu
+////////////////////////////////////////////////////////////////////////
+
+#include "Geant4/G4LossTableManager.hh"
+#include "Geant4/G4EmSaturation.hh"
+#include "Geant4/G4ParticleTypes.hh"
+
+#include "larcore/CoreUtils/ServiceUtil.h"
+#include "larsim/LegacyLArG4/ISCalculationCorrelated.h"
+#include "lardata/DetectorInfoServices/LArPropertiesService.h"
+#include "lardata/DetectorInfoServices/DetectorPropertiesService.h"
+#include "larsim/Simulation/LArG4Parameters.h"
+#include "larsim/Simulation/LArVoxelCalculator.h"
+
+#include "messagefacility/MessageLogger/MessageLogger.h"
+#include "cetlib_except/exception.h"
+
+
+namespace larg4{
+
+  //----------------------------------------------------------------------------
+  ISCalculationCorrelated::ISCalculationCorrelated(CLHEP::HepRandomEngine&)
+  {
+  }
+
+  //----------------------------------------------------------------------------
+  ISCalculationCorrelated::~ISCalculationCorrelated()
+  {
+  }
+
+  //----------------------------------------------------------------------------
+  void ISCalculationCorrelated::Initialize()
+  {
+    std::cout << "LegacyLArG4/ISCalculationCorrelated Initialize." << std::endl;
+    art::ServiceHandle<sim::LArG4Parameters const> lgpHandle;
+    const detinfo::LArProperties* larp = lar::providerFrom<detinfo::LArPropertiesService>();
+    const detinfo::DetectorProperties* detprop = lar::providerFrom<detinfo::DetectorPropertiesService>();
+
+    double density        = detprop->Density(detprop->Temperature());
+    fEfield               = detprop->Efield();
+
+    // ionization work function
+    fWion                 = 1./lgpHandle->GeVToElectrons() * 1e3; // MeV
+
+    // ion+excitation work function (\todo: get from LArG4Parameters or LArProperties?)
+    fWph                  = 19.5 * 1e-6; // MeV
+
+    // the recombination coefficient is in g/(MeVcm^2), but
+    // we report energy depositions in MeV/cm, need to divide
+    // Recombk from the LArG4Parameters service by the density
+    // of the argon we got above.
+    fRecombA             = lgpHandle->RecombA();
+    fRecombk             = lgpHandle->Recombk()/density;
+    fModBoxA             = lgpHandle->ModBoxA();
+    fModBoxB             = lgpHandle->ModBoxB()/density;
+    fUseModBoxRecomb     = lgpHandle->UseModBoxRecomb();
+
+    // determine the step size using the voxel sizes
+    art::ServiceHandle<sim::LArVoxelCalculator const> lvc;
+    double maxsize = std::max(lvc->VoxelSizeX(), std::max(lvc->VoxelSizeY(), lvc->VoxelSizeZ())) * CLHEP::cm;
+
+    fStepSize = 0.1 * maxsize;
+
+    return;
+  }
+
+  //----------------------------------------------------------------------------
+  // fNumIonElectrons returns a value that is not corrected for life time effects
+  void ISCalculationCorrelated::Reset()
+  {
+    fEnergyDeposit   = 0.;
+    fNumScintPhotons = 0.;
+    fNumIonElectrons = 0.;
```

Comment:
The `fEnergyDeposit` and `fNumIonElectrons` data members are only used within the `calculateIonizationAndScintillation` function and should therefore be removed as class data members.  It is much better to define these variables only in the function where they are used.